README.md
# Bitrise Workflow Editor

## How to install & use the Workflow Editor on your Mac/Linux

1. Install [Go](https://golang.org) `brew install go`(on macOS)
1. Install the latest [Bitrise CLI](https://www.bitrise.io/cli) - it's a single binary command line tool
1. Run `bitrise setup` just to be sure everything's prepared
1. `cd` into a directory where you have your `bitrise.yml`, and run: `bitrise :workflow-editor`

That's all. The Workflow Editor is now part of the Bitrise CLI core plugins, so you don't have to install it manually.

To upgrade to the latest version of the Workflow Editor run:

```
bitrise plugin update workflow-editor
```

*Join the Workflow Editor's discussion at: [https://discuss.bitrise.io/t/workflow-editor-v2-offline-workflow-editor/39](https://discuss.bitrise.io/t/workflow-editor-v2-offline-workflow-editor/39)*


## Install requirements

Workflow editor uses webpack for static asset compilation and asset bundling. For transformation we need to use some rails related transformation hence it also uses bundler to install ruby dependencies. In addition it uses karma and jasmine for frontend tests execution so it needs `node` and `npm` installed to get the dependencies for testing and also production.

Finally the local executable is written in GO. so you need to have go set up as well and dependencies.

Something like this:

```bash
bundle install
npm install
```

## Development

### Build a stand-alone binary, with embedded resources

```
bitrise run go-install
```

### Run in development mode

1. __DOCKER__: In the Workflow Editor's directory, run `bitrise run up`. This utilizes `docker-compose` behind the scenes. // __LOCAL__: You need to start the api `bitrise run up-api` and webpack dev server  `bitrise run up-dev-server`.
1. In your browser, you can reach the Workflow Editor on `localhost:1234`. Be aware that you usually have to wait a while until dev server starts up.
1. By default, the Workflow Editor will open the bitrise.yml and .bitrise.secrets.yml found in this folder. For testing purposes, you probably want to be able to edit custom files. This can be achieved by setting the `TEST_BITRISE_CONFIG_PATH` and `TEST_BITRISE_SECRETS_PATH` environment variables with the path pointing to the custom files' paths.

### Run client tests

Use `npm test` for a single test run.
If you only iterate on tests, you can also use `npm run karma` as it skips transpilation and the transpilation and run the tests on an already transpiled JS. (faster)

# Contributing

This project is using squash & merge model, feel free to have as many commits as you like but at the end the work will end up on master as a single commit.

**IMPORTANT** Please make sure every commit on the master is released, there are no expections. Someone from bitrise should be able to assist with that!
If you are planning to break down the tasks into multiple master commits (aka milestones) please use a long-living feature branch and fork that to create additional branches.

## New version release

- Generate a GitHub personal access token for your user with `repo` access (one who has rights to create releases on the repository) - you can generate one here: https://github.com/settings/tokens
  ![](https://user-images.githubusercontent.com/15610939/75050804-0361ee80-54cd-11ea-85c6-0c89974051b7.png)
- Generate a Discuss API key or use the common key for _All Users_.
  You need to be a Discourse admin for this, then you can generate an API key for yourself at: https://discuss.bitrise.io/admin/api/keys
- Make sure you're on the latest `master`
- If new release requires Bitrise CLI to be updated, in `bitrise-plugin.yml` change `min_version` requirement of the `bitrise` tool to the required CLI version
- Optional: set the following secrets: $GITHUB_RELEASE_API_TOKEN, $GITHUB_USERNAME, $DISCUSS_API_KEY, $DISCUSS_USERNAME
- Call `bitrise run create-release`
- During the build you will need to specify a new version number, and if you did not specify any of the secrets above, you will need to specify those as well.
- After the build has finished, close the related GitHub issues, and milestones if the issues were assigned to any.

## Testing if version release works, without actually releasing

- In bitrise.yml, create a workflow e. g. `test-release`
- From the `create-release` workflow, copy-paste the *GitHub release* and *Create Discuss topic* steps.
- In the GitHub release step, remove the `files_to_upload` input, set the `$NEW_RELEASE_VERSION` everywhere to something arbitrary, same for the `body`, and **most importantly set `draft: 'yes'`**
- In the Create Discuss topic step, **change the `DISCUSS_CHANGELOG_CATEGORY_ID` to the ID of one our discuss.bitrise.io's internal channels' ID** (you can find an ID using the Discourse API with a cURL request) so that it is only visible to us; also change the `title` and the `raw` parameter to something arbitrary.
- After the test release process, don't forget to delete the draft release and the internal changelog topic.

## Contribution guide

...Process TBD...

### Javascript

For production code we still use ES5 standard as we do not do any transpilation during the build (only minification). For tests you are safe to use whatever standards jsdom executes (ES6 supported).


CHANGELOG.md
## Changelog (Current version: 1.3.1)

-----------------

## 1.3.1 (2020 Mar 02)

### Release Notes

* Fix search deprecated steps, List steps by ID not CVS (#455)
* change terser config to preserve licence (#454)
* fix master deploy


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.3.1
```

To update the already installed plugin to the latest version:

```
bitrise plugin update workflow-editor
```

That's all, you're ready to go!

To run the workflow editor plugin call:

```
bitrise :workflow-editor
```

### Release Commits - 1.3.0 -> 1.3.1

* [6c0b75b] Simon Márton - Fix search deprecated steps, List steps by ID not CVS (#455) (2020 Mar 02)
* [b937ffb] Adam Pelle - change terser config to preserve licence (#454) (2020 Mar 02)
* [8dfef43] Adam Pelle - fix master deploy (2020 Mar 02)


## 1.3.0 (2020 Mar 02)

### Release Notes

* [TOOL-1197] step version lock release (#446)
* Remove some unnecessary AJAX request duplications [WEB-2015] (#445)
* update go dependencies (#452)
* Import js-yaml (#453)
* Import js-yaml (#453)
* Separate GH and Discuss release WF (#448)


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.3.0
```

To update the already installed plugin to the latest version:

```
bitrise plugin update workflow-editor
```

That's all, you're ready to go!

To run the workflow editor plugin call:

```
bitrise :workflow-editor
```

### Release Commits - 1.2.1 -> 1.3.0

* [64e0149] Adam Pelle - [TOOL-1197] step version lock release (#446) (2020 Mar 02)
* [8dbb6de] erosdome - Remove some unnecessary AJAX request duplications [WEB-2015] (#445) (2020 Mar 02)
* [6a29b11] Adam Pelle - update go dependencies (#452) (2020 Feb 28)
* [8a2ac46] Simon Márton - Import js-yaml (#453) (2020 Feb 27)
* [8a64976] Simon Márton - Import js-yaml (#453) (2020 Feb 27)
* [87921b8] Simon Márton - Separate GH and Discuss release WF (#448) (2020 Feb 27)


## 1.2.1 (2020 Feb 27)

### Release Notes

* Add Workflows tab specification for Workflow control bar, Step list, Step details, Step Inputs (part of it) (#450)
* Fix steplib issues (#451)


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.2.1
```

To update the already installed plugin to the latest version:

```
bitrise plugin update workflow-editor
```

That's all, you're ready to go!

To run the workflow editor plugin call:

```
bitrise :workflow-editor
```

### Release Commits - 1.2.0 -> 1.2.1

* [7cae6ce] erosdome - Add Workflows tab specification for Workflow control bar, Step list, Step details, Step Inputs (part of it) (#450) (2020 Feb 26)
* [bfa7c87] Simon Márton - Fix steplib issues (#451) (2020 Feb 26)


## 1.2.0 (2020 Feb 21)

### Release Notes

* Port StepLib search service to TS (#447)
* Reference Workflow Editor inside Go path in API dev mode (#444)
* [WEB-3399] Use steplib search service for fuzzy search  (#443)
* Fix unnecessary polling on Code Signing tab
* [WEB-3398] Load selected project type in step selector by default (#441)
* [WEB-3397] Integrate and use StepLib Search NPM package (#437)


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.2.0
```

To update the already installed plugin to the latest version:

```
bitrise plugin update workflow-editor
```

That's all, you're ready to go!

To run the workflow editor plugin call:

```
bitrise :workflow-editor
```

### Release Commits - 1.1.74 -> 1.2.0

* [d488da0] Simon Márton - Port StepLib search service to TS (#447) (2020 Feb 21)
* [655323a] erosdome - Reference Workflow Editor inside Go path in API dev mode (#444) (2020 Feb 20)
* [cb964c1] Simon Márton - [WEB-3399] Use steplib search service for fuzzy search  (#443) (2020 Feb 20)
* [c1cb6de] erosdome - Fix unnecessary polling on Code Signing tab (2020 Feb 20)
* [fcef8a1] Simon Márton - [WEB-3398] Load selected project type in step selector by default (#441) (2020 Feb 18)
* [9a298e1] Simon Márton - [WEB-3397] Integrate and use StepLib Search NPM package (#437) (2020 Feb 18)


## 1.1.74 (2020 Feb 17)

### Release Notes

* Fix flex bug (#439)
* [WEB-3293] Handle null response at YML validation (#440)
* master deploy fixes (#436)


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.1.74
```

To update the already installed plugin to the latest version:

```
bitrise plugin update workflow-editor
```

That's all, you're ready to go!

To run the workflow editor plugin call:

```
bitrise :workflow-editor
```

### Release Commits - 1.1.73 -> 1.1.74

* [4543173] erosdome - Fix flex bug (#439) (2020 Feb 17)
* [efc4f51] Gergely Békési - [WEB-3293] Handle null response at YML validation (#440) (2020 Feb 17)
* [11ab7a6] Adam Pelle - master deploy fixes (#436) (2020 Feb 14)


## 1.1.73 (2020 Feb 13)

### Release Notes

* [WEB-3426] invalid version handling (#430)
* Fix webpack deploy (#435)
* Revert "version bump (1.1.72 -> 1.1.73) (#433)" (#434)
* remove unnecessary check (#431)
* version bump (1.1.72 -> 1.1.73) (#433)
* godeps update to pull new versions of bitrise and stepman (#432)
* Webpack integration (#429)


### Install or upgrade

To install this version, run the following commands (in a bash shell):

```
bitrise plugin install --source https://github.com/bitrise-io/bitrise-workflow-editor.git --version 1.1.73
```

To update the already installed plugin to the latest version:

```

karma.conf.js
// Karma configuration
// Generated on Wed Jan 15 2020 13:26:53 GMT+0100 (Central European Standard Time)

// verbose logging
process.on('infrastructure_error', (error) => {
  console.error('infrastructure_error', error);
});


module.exports = (config) => {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: 'build',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'javascripts/angular_cache.js',
      '../node_modules/angular-mocks/angular-mocks.js',
      'javascripts/*.js',
      '../spec/javascripts/**/*.spec.js'
    ],


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'javascripts/main.js': ['coverage']
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['spec', 'coverage'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['jsdom'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: 4,

    coverageReporter: {
      type: 'text-summary'
    }
  })
}
